tools/analysis_tools/visualize_tsne.py

```diff
@@ -184,64 +183,47 @@ def main():
         worker_init_fn=init_fn,
         **extract_dataloader_cfg)
 
-    # build the model
-    # get backbone out_indices from pool_cfg in tsne_dataset_config
-    cfg.model.backbone.out_indices = dataset_cfg.pool_cfg.get(
-        'in_indices', [4])
-    logger.info(
-        f'The output indices of backbone is {cfg.model.backbone.out_indices}.')
-    model = MODELS.build(cfg.model)
-    model.init_weights()
+    results = dict()
+    features = []
+    labels = []
+    progress_bar = mmengine.ProgressBar(len(tsne_dataloader))
+    for _, data in enumerate(tsne_dataloader):
+        with torch.no_grad():
+            # preprocess data
+            data = model.data_preprocessor(data)
+            batch_inputs, batch_data_samples = \
+                data['inputs'], data['data_samples']
 
-    # model is determined in this priority: init_cfg > checkpoint > random
-    if hasattr(cfg.model.backbone, 'init_cfg'):
-        if getattr(cfg.model.backbone.init_cfg, 'type', None) == 'Pretrained':
-            logger.info(
-                f'Use pretrained model: '
-                f'{cfg.model.backbone.init_cfg.checkpoint} to extract features'
-            )
-    elif args.checkpoint is not None:
-        logger.info(f'Use checkpoint: {args.checkpoint} to extract features')
-        load_checkpoint(model, args.checkpoint, map_location='cpu')
-    else:
-        logger.info('No pretrained or checkpoint is given, use random init.')
+            # extract backbone features
+            batch_features = model.extract_feat(
+                batch_inputs, stage=args.vis_stage)
 
-    if torch.cuda.is_available():
-        model = model.cuda()
+            # post process
+            if args.vis_stage == 'backbone':
+                batch_features = [
+                    F.adaptive_avg_pool2d(inputs, 1).squeeze()
```

Comment:
For ViT, global average pool is problematic, since outputs from the backbone contain the cls token and all other tokens. Better provide an option for users to choose the cls token or the average of all other tokens as the final prediction features.

Reply:
The default setting in config uses averaged cls_token and removed cls_token.
But it is correct to add this option.